src/Model.php

```diff
@@ -171,6 +172,8 @@ class Model implements \ArrayAccess, \IteratorAggregate
      */
     public $conditions = [];
 
+    protected $scope;
```

Comment:
is there support for HAVING .. ?

Reply:
no

Reply:
this is quite important point at least to be easily supported in the future and we should probably use HAVING automatically where needed

Reply:
Let me have an example HAVING query use case and I will work to implement it in scopes

Reply:
Table product
varchar name
varchar category
int price;

insert into product (name, price) values
('apple', 'fruit', 10)
('potato', 'vegetable', 2)
('mango', 'fruit', 30)

SELECT category, avg(price) as avg_price
FROM product
GROUP BY category
HAVING avg_price > 15

should not fail (with SQL `WHERE` is will) and return `fruit, 20`

Reply:
HAVING is used only when grouping and that can be done by GroupModel. See https://github.com/atk4/report/blob/develop/src/GroupModel.php#L319.
So maybe Having should be used automatically in GroupModel but normally in standard model it probably have no use.

Reply:
I do not think so - any expr. field with non trivial expression (ie. pure field name) needs having:

Example:
```
SELECT CONCAT(name, 'dummy') as x FROM `test` where x
```
throws an error: `#1054 - Unknown column 'x' in 'where clause'`

Reply:
Yes and no. In this case you *can* use `having x` instead of `where x`, but need to keep in mind that `having` works only after data is selected. So if you use `having` then DB engine first will select all data and only then start filtering it. That's slower and requires more resources from DB server than using `where`. `Where` instead filters data while it's selected, so resulting dataset is smaller and performance better.
So instead of using `having` you should always use `where` if possible and it's always possible if there are no grouping.
In your testcase above generated SQL should be `SELECT CONCAT(name, 'dummy') as x FROM `test` where CONCAT(name, 'dummy')` and atk4/data helps with that automatically if concat is defined as expression field in model.

Reply:
yes and no

I think queries like select (select id from xx limit 1) where (select id from xx limit 1) are not guaranteed to be stable (because id can be different each time if there is no forced order)

I think there are also other major issues atk4 only assume instead of can rely on. But to keep this thread simple, I proposed to isolate Scope idea (ie. Scope class to define nested where conditions) from anything else and solve any freecoolin stuff later /in another PR.